### PR TITLE
wallet: Fix non-determinism in ParseHDKeypath(...). Avoid using an uninitialized variable in path calculation.

### PR DESCRIPTION
**Pull request by practicalswift**
Add error handling. Check return value of `ParseUInt32(...)` in `ParseHDKeypath(...)`.

`ParseUInt32(...)` returns `false` if the entire string could not be parsed or when an overflow or underflow occurred. In such case the uninitialized variable `number` would be used in the calculation of `path` (prior to this commit).

An example key path triggering this is `m/0/4294967296`:

```
  ParseHDKeypath("m/0/4294967296", keypath);
```

`4294967296` is `1` + `0xFFFFFFFF` (`uint32_t` max: `4294967295`).

Introduced in a4b06fb42eb0ad94e562ca839391b57e69285136 which was merged into `master` 14 hours ago as part of #13557 ("BIP 174 PSBT Serializations and RPCs").

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Would be nice if this was covered by a test case, no?

**Comment by practicalswift** [comment]
@MarcoFalke Thanks for the quick review. I've now added some tests.

These are the now added tests that fails before the fix commit is applied and passes after the fix commit is applied:

```
wallet/test/psbt_wallet_tests.cpp(79): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("///////////////////////////", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(82): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("//////////////////////////'/", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(85): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("1///////////////////////////", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(88): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("1/'//////////////////////////", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(109): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("4294967296", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(133): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("m/1/1/111111111111111111111111111111111111111111111111111111111111111111111111111111111111", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(139): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("m/1//", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(142): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("m/0/4294967296", keypath) has failed
wallet/test/psbt_wallet_tests.cpp(145): error: in "psbt_wallet_tests/parse_hd_keypath": check !ParseHDKeypath("m/4294967296", keypath) has failed
```

File: src/wallet/rpcwallet.h [hunk 1]
```diff
@@ -31,4 +31,5 @@ bool EnsureWalletIsAvailable(CWallet *, bool avoidException);
 UniValue getaddressinfo(const JSONRPCRequest& request);
 UniValue signrawtransactionwithwallet(const JSONRPCRequest& request);
 bool FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const CTransaction* txConst, int sighash_type = 1, bool sign = true, bool bip32derivs = false);
+bool ParseHDKeypath(std::string keypath_str, std::vector<uint32_t>& keypath);
```
Comment (Empact) [inline src/wallet/rpcwallet.h]: Rather than declare it in the header, you could declare it `extern` in the test file.

**Comment by achow101** [comment]
utACK 722326389989757c58879feb74cc2480bd58bdfc

**Comment by MarcoFalke** [comment]
utACK 7223263

**Comment by MarcoFalke** [comment]
utACK 27ee53c1aedae4f27458537c61804b6fef34ce3c

**Comment by sipa** [comment]
utACK 27ee53c1aedae4f27458537c61804b6fef34ce3c

**Comment by practicalswift** [comment]
@achow101 @MarcoFalke @sipa As reviewers of this bug fix you might be interested in reviewing PR #13815 which adds annotations (C++17-style `[[nodiscard]]` or `__attribute__((warn_unused_result))` depending on availability) which could help us guard against discarded return values going forward :-)

**Pull request merged**